src/CareTogether.Core/Utilities/AppendBlobMultitenantEventLog.cs

```diff
@@ -0,0 +1,130 @@
+
+            var eventStream = new MemoryStream(Encoding.UTF8.GetBytes(eventJson));
+
+            try
+            {
+                var appendResult = await logSegmentBlob.AppendBlockAsync(eventStream);
+
+                if(appendResult.Value.BlobCommittedBlockCount == expectedSequenceNumber)
+                {
+                    // if the append was successful then the number of tenant events needs to increment
+                    var updateSuccessful = _tenantEventCounts.TryUpdate((organizationId,locationId), numberOfTenantEvents+1, numberOfTenantEvents);
+
+                    if(updateSuccessful)
+                    {
+                        return new Success();
+                    } else
+                    {
+                        throw new Exception($"Tried to increment the number of tenant events but the current value for tenant {organizationId} and location {locationId} " +
+                            $"did not match the value we expected. This probably means another process updated the dictionary.");
+                    }
+                } else
+                {
+                    throw new Exception($"The append block number {appendResult.Value.BlobCommittedBlockCount} did not match the expected block number {expectedSequenceNumber}");
+                }
+            }
+            catch (Exception e)
+            {
+                System.Diagnostics.Debug.WriteLine($"There was an issue appending to block {expectedSequenceNumber} in {locationId}/{_logType}/{currentBlobNumber:D5}.ndjson" +
+                    $" under tenant {organizationId}: "+e.Message);
+                return new Error();
+            }
+        }
+
+        public async IAsyncEnumerable<(T DomainEvent, long SequenceNumber)> GetAllEventsAsync(Guid organizationId, Guid locationId)
+        {
+            var tenantContainer = await createContainerIfNotExists(organizationId);
+
+            long eventSequenceNumber = 0;
+
+            await foreach (BlobItem blob in tenantContainer.GetBlobsAsync(BlobTraits.None, BlobStates.None, $"{locationId}/{_logType}"))
+            {
+                var appendBlob = tenantContainer.GetAppendBlobClient(blob.Name);
+   
+                var eventStream = await appendBlob.OpenReadAsync();
+
+                string eventString = new StreamReader(eventStream).ReadToEnd();
+
+                using (StringReader reader = new StringReader(eventString))
+                {
+                    string line;
+
+                    while ((line = reader.ReadLine()) != null)
+                    {
+                        eventSequenceNumber++;
+
+                        T item = JsonConvert.DeserializeObject<T>(line);
+
+                        yield return (item, eventSequenceNumber);
+                    }
+                }
+            }
+
+            _tenantEventCounts.TryAdd((organizationId, locationId), (int)eventSequenceNumber);
```

Comment:
I think `TryAdd` is incorrect here. It'll only succeed if the key does not already exist ([docs](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.tryadd?view=net-5.0)). You can just set the entry with an indexer:
```suggestion
            _tenantEventCounts[(organizationId, locationId)] = (int)eventSequenceNumber;
```

Reply:
As you pointed out, we can just remove this altogether.